File: Java/net.alcetech.UniversalEditor.Plugins.Core/src/net/alcetech/UniversalEditor/ObjectModels/Markup/MarkupTagElement.java
package net.alcetech.UniversalEditor.ObjectModels.Markup;

import java.util.ArrayList;

import net.alcetech.UniversalEditor.ObjectModels.Markup.*;

public class MarkupTagElement extends MarkupElement
{
	public MarkupTagElement(String fullName)
	{
		super(fullName);
	}
	public MarkupTagElement(String fullName, String value)
	{
		super(fullName, value);
	}
	
	private ArrayList<MarkupAttribute> _attributes = new ArrayList<MarkupAttribute>();
	public void addAttribute(MarkupAttribute item) {
		_attributes.add(item);
	}
	public MarkupAttribute addAttribute(String name) {
		return addAttribute(name, "");
	}
	public MarkupAttribute addAttribute(String name, String value) {
		MarkupAttribute att = new MarkupAttribute(name, value);
		addAttribute(att);
		return att;
	}
	public MarkupAttribute[] getAttributes() {
		MarkupAttribute[] array = new MarkupAttribute[_attributes.size()];
		_attributes.toArray(array);
		return array;
	}
	public MarkupAttribute getAttribute(String name) {
		MarkupAttribute[] atts = this.getAttributes();
		for (int i = 0; i < atts.length; i++) {
			if (atts[i].getName().equals(name)) {
				return atts[i];
			}
		}
		return null;
	}

	private ArrayList<MarkupElement> _elements = new ArrayList<MarkupElement>();
	public void addElement(MarkupElement item) {
		item.setParentElement(this);
		_elements.add(item);
	}
	public MarkupElement[] getElements() {
		MarkupElement[] elements = new MarkupElement[_elements.size()];
		_elements.toArray(elements);
		return elements;
	}
	
	public MarkupTagElement getTagByTagName(String name) {
		MarkupElement[] elements = getElements();
		for (int i = 0; i < elements.length; i++) {
			if (elements[i].getName().equals(name)) {
				return (MarkupTagElement)elements[i];
			}
		}
		return null;
	}
	public MarkupTagElement[] getTags() {
		ArrayList<MarkupTagElement> list = new ArrayList<MarkupTagElement>();
		MarkupElement[] elements = getElements();
		for (int i = 0; i < elements.length; i++) {
			if (elements[i].getClass().isAssignableFrom(MarkupTagElement.class)) {
				list.add((MarkupTagElement)elements[i]);
			}
		}
		
		MarkupTagElement[] array = new MarkupTagElement[list.size()];
		list.toArray(array);
		return array;
	}

}

File: CSharp/Libraries/UniversalEditor.Essential/ObjectModels/Markup/MarkupTagElement.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace UniversalEditor.ObjectModels.Markup
{
	public class MarkupTagElement : MarkupContainerElement
	{
		private MarkupAttribute.MarkupAttributeCollection mvarAttributes = new MarkupAttribute.MarkupAttributeCollection();
		public MarkupAttribute.MarkupAttributeCollection Attributes
		{
			get
			{
				return this.mvarAttributes;
			}
		}
		public override object Clone()
		{
			MarkupTagElement clone = new MarkupTagElement();
			foreach (MarkupElement el in base.Elements)
			{
				clone.Elements.Add(el.Clone() as MarkupElement);
			}
            foreach (MarkupAttribute att in mvarAttributes)
            {
                clone.Attributes.Add(att.Clone() as MarkupAttribute);
            }
			clone.Name = base.Name;
			clone.Namespace = base.Namespace;
			clone.Value = base.Value;
			return clone;
		}

        public MarkupAttribute FindAttribute(string Name)
        {
            return FindAttribute(Name, (MarkupAttribute)null);
        }
        public string FindAttribute(string Name, string DefaultValue)
        {
            MarkupAttribute att = FindAttribute(Name);
            if (att == null) return DefaultValue;
            return att.Value;
        }
        public MarkupAttribute FindAttribute(string Name, MarkupAttribute defaultValue)
        {
            MarkupAttribute att = mvarAttributes[Name];
            if (att == null) return defaultValue;
            return att;
        }

		public MarkupTagElement[] GetElementsByTagName(string tagName)
		{
			List<MarkupTagElement> list = new List<MarkupTagElement>();
			for (int i = 0; i < Elements.Count; i++)
			{
				if ((Elements[i] as MarkupTagElement)?.FullName == tagName)
					list.Add(Elements[i] as MarkupTagElement);
			}
			return list.ToArray();
		}
	}
}
